Added a basic monitor for providing etcd version related info

shyamjvs opened the pull request:
Fixes #41071 

This tool scrapes metrics partly from etcd's /version and /metrics endpoints and partly using etcdctl and exposes them as prometheus metrics at `http://localhost:9101/metrics` endpoint on the master. Here is a summary of the metrics it exposes (self-explanatory from the code):
-        etcdVersionFetchCount   = prometheus.NewCounterVec(
                prometheus.CounterOpts{
                        Namespace: "etcd",
                        Name: "version_info_fetch_count",
                        Help: "Number of times etcd's version info was fetched, labeled by etcd's server binary and cluster version",
                },
                []string{"serverversion", "clusterversion"})
-         etcdGRPCRequestsTotal   = prometheus.NewCounterVec(
                prometheus.CounterOpts{
                        Namespace: namespace,
                        Name: "grpc_requests_total",
                        Help: "Counter of received grpc requests, labeled by grpc method and grpc service names",
                },
                []string{"grpc_method", "grpc_service"})

For further info on how to run this as a binary/docker-container/kubernetes-pod and checking the metrics, have a look at the README.md file.

cc @fgrzadkowski @wojtek-t @piosz

shyamjvs committed:
Added a basic monitor for watching etcd version and size related info

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41080)
<!-- Reviewable:end -->

piosz commented:
Please split your PR into 2 commits:
- one with real changes
- one with godeps.

Also is core Kubernetes repo a suitable place for such component? How about using contrib?

shyamjvs commented:
@piosz Split it into 2 commits. PTAL now.

liggitt commented:
why does this need to be in the kubernetes repo?

shyamjvs commented:
@liggitt @piosz contrib repo sgtm for placing this.
Will wait for @wojtek-t to give his opinion.

shyamjvs commented:
@k8s-bot unit test this
@k8s-bot verify test this
@k8s-bot cvm gce e2e test this

shyamjvs commented:
@k8s-bot verify test this

shyamjvs commented:
@piosz @liggitt Seems like the plan is to eventually move out all code from the contrib repo. So placing this code there might not be a good idea.
For now, cluster/images seems to be a good place to have this code.

mikedanese commented:
Please don't add a new vendor directory. Just use the one in the root.

shyamjvs commented:
@mikedanese Thanks for pointing this out. I have removed Godeps and vendor code for now. Will add it in a separate commit once this PR gets LGTM'ed, otherwise it could be a significant overhead to update Godep each time as the process is quite time-consuming.

liggitt commented:
Pretty sure the binary etcdctl shouldn't be checked in. How many additional godeps are needed for this? I still would push on this not being in the core repo.

shyamjvs commented:
@wojtek-t Added `etcd_grpc_requests_total` metric as well, so now all metrics for version related monitoring we want are available under a single roof.

@liggitt Even I feel that having this binary checked in to the repo is not a good option. We can alternatively:
- Use etcdctl directly by importing it's code (but I'm not very inclined towards this solution due to additional Godep handling burden for etcdctl too)
- Get (install) these binaries into the bin/ folder as part of building the docker image inside the Makefile (this seems much cleaner).

@wojtek-t What's your opinion on this?

shyamjvs commented:
@liggitt I just checked and it turns out this tool is pretty simple and all of its dependencies (except "github.com/prometheus/common/log", which can also be replaced with some other logging package) are already vendored. So if we follow the second solution above, we shouldn't be needing any (/ much) godep changes :)

wojtek-t commented on cluster/images/etcd-version-monitor/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Please remove this licence. This is part of Kubernetes and should be under its licence.

shyamjvs replied:
Done. Sorry for the blooper.

wojtek-t commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,43 @@
+
+# Build the binary for etcd-monitor.
+build: etcd-version-monitor.go
+	GOOS=linux CGO_ENABLED=0 go build -o ./bin/etcd-version-monitor etcd-version-monitor.go
```
This depends on Go version of the person that is building the image - this is bad.
Please use our standard mechanism of building in Docker. For the example you can see here:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/Makefile#L59

shyamjvs replied:
Thanks for pointing this out, learnt quite a lot in the process of fixing this :)

wojtek-t commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,43 @@
+# Remove the binary.
+clean:
+	rm -f bin/etcd-version-monitor
+
```
Remove empty line at the end.

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/README.md:
```diff
@@ -0,0 +1,30 @@
+# etcd-version-monitor
+
+This is a tool for exporting metrics exposed by etcd endpoints that are not
+in prometheus metrics format. The exported metrics would be in prometheus
```
This sentence is no longer true - we are exposing some that are already in prometheus format.

Also, this description is in general not accurate. The purpose of this tool is to expose etcd metrics that might be interesting from monitoring purposes in a prometheus format (no matter how do we get them).

shyamjvs replied:
Fixed.

wojtek-t commented on cluster/images/etcd-version-monitor/README.md:
```diff
@@ -0,0 +1,30 @@
+
+To run this tool as a docker container:
+- make build-image
+- docker run --net=host -i -t gcr.io/google_containers/etcd_version_monitor:test /bin/etcd-version-monitor
```
The TAG shouldn't be test when we commit it.

shyamjvs replied:
Yes. Used test till now because I was mostly experimenting. Would "0.1" be an apt tag?

wojtek-t commented on cluster/images/etcd-version-monitor/README.md:
```diff
@@ -0,0 +1,30 @@
+To run this tool as a docker container:
+- make build-image
+- docker run --net=host -i -t gcr.io/google_containers/etcd_version_monitor:test /bin/etcd-version-monitor
+<br /> Optionally, change the 'IMAGE' variable inside the Makefile and run 'make push_image'
```
I don't think it brings any value.

shyamjvs replied:
Removed the point about building the image with a different name. However, wrote the note about running directly using docker for people who might want to use this but not in the context of k8s.
Do you suggest removing this point?

wojtek-t commented on cluster/images/etcd-version-monitor/README.md:
```diff
@@ -0,0 +1,30 @@
+- Place the 'etcd-version-monitor.yaml' in the manifests directory of kubelet on
+  the master machine.
+
+*Note*: In all the 3 modes above, you have to ensure that this tool runs on
```
It not in all 3 modes - you should always do that, no matter how you decide to run it (e.g. under supervisor).

shyamjvs replied:
Fixed.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+/*
```
Couldn't comment above, so commenting here.

Where those binaries come from? It's not a good practise to submit binaries to the repository.

Instead of that, you should download them as part of building the image.

shyamjvs replied:
Yes, I felt the same too. Changed it to download etcdctl (the only extra binary we need) as part of the makefile.

shyamjvs replied:
Done.

lavalamp replied:
I do not see it downloading etcdctl? I'd greatly prefer if this did not download or use etcdctl, so that we don't have to worry about the version being out of sync.

wojtek-t replied:
Yes - this was removed. So it no longer depends on etcdctl.

wojtek-t commented on cluster/images/etcd-version-monitor/README.md:
```diff
@@ -0,0 +1,30 @@
+
+To run this tool as a binary:
+- make
+- ./bin/etcd-version-monitor
```
This will not work if the other binaries are not in correct places. I wouldn't recommend this method.

shyamjvs replied:
Yes I agree. Removed it.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+
+// Initialize the prometheus instrumentation and client related flags.
+var (
+	listenAddress				string
```
Please make the spaces reasonable (as in Go styleguide)

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&listenAddress, "listen-address", ":9101", "Address to listen on for serving prometheus formatted metrics")
+	fs.StringVar(&metricsPath, "metrics-path", "/metrics", "Path under which prometheus formatted metrics are to be served")
```
s/formatted//

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&listenAddress, "listen-address", ":9101", "Address to listen on for serving prometheus formatted metrics")
```
s/formatted//

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name: "version_info_fetch_count",
+			Help: "Number of times etcd's version info was fetched, labeled by etcd's server binary and cluster version",
```
I completely don't understand this. What this metric exposes?

We wanted to expose "binary version" (e.g. "3.0.14" or "2.2.1").

shyamjvs replied:
Metric values can only be floats. So capturing such variables (like version) has to be done through labels.
And for the value of this metric, it seemed quite reasonable to have count of the no. of times we fetched the version info.
We can proabably have something else (like last fetch timestamp (?))

wojtek-t replied:
@fgrzadkowski @piosz - any thoughts what would be the best one?

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name: "versioned_data_size",
+			Help: "Size of etcd data in a given backend version (>=3), labeled by backend version, unit for size and etcdctl version used",
```
Labeld by "unit"? What do you mean? It doesn't sound like a good way to do this.

shyamjvs replied:
Changed it "unit_for_size". It's basically KB, MB, etc.
Would you rather suggest converting everything to bytes or KB?

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+			Namespace: namespace,
+			Name: "versioned_data_size",
+			Help: "Size of etcd data in a given backend version (>=3), labeled by backend version, unit for size and etcdctl version used",
+		},
```
Also, why do you need to label it with "etcdctl version" ?

shyamjvs replied:
It's not required anymore, so removed it. Previously I did it just for info as different people could be using different version of it, but now I standardized it to 3.0.14 as part of makefile.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+	// Create the get request for the etcd endpoint.
+	req, err := http.NewRequest("GET", etcdVersionScrapeURI, nil)
+	if err != nil {
+		log.Fatal("NewRequest: ", err)
```
Don't do Fatals - log errors.

Also, this whole "getVersion" method should return error.

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		client := &http.Client{}
+		resp, err := client.Do(req)
+		if err != nil {
+			log.Fatal("Do: ", err)
```
remove

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		return
+	}
+
+	for {
```
It would be a better pattern for this function to just to a single request, and then do the loop over time in a separate function.

shyamjvs replied:
Done. Thanks for this suggestion, stuff looks much cleaner now.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		// Obtain EtcdVersion from the JSON response.
+		var version EtcdVersion
+		if err := json.NewDecoder(resp.Body).Decode(&version); err != nil {
+			log.Fatal("JSON decoder for etcd version: ", err)
```
don't fatal

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+			time.Sleep(scrapeTimeout)
+			continue
+		}
+		resp.Body.Close()
```
this should be deferred previously (but we need to have this function to handle a single request).

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		env := os.Environ()
+		env = append(env, fmt.Sprintf("ETCDCTL_API=3"))
+		cmd.Env = env
+		cmdOut, _ := cmd.StdoutPipe()
```
Why you are ignoring error?

shyamjvs replied:
Fixed.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+			Help: "Number of times etcd's version info was fetched, labeled by etcd's server binary and cluster version",
+		},
+		[]string{"server_version", "cluster_version"})
+	etcdVersionedDataSize 	= prometheus.NewGaugeVec(
```
I think that we generally agreed, to not expose this metric for now, as it may be too misleading.

wojtek-t replied:
@fgrzadkowski - can you confirm?

shyamjvs replied:
IIUC this value would be generally helpful to indicate that things are moving in v3. Though it might not be the same as size of the data (because it's of allocated pages), in time it should (as pointed out by @xiang90) grow linearly with the size of the data.

wojtek-t replied:
We don't want to expose it for now, because:
1. we agreed this metric is not super useful (it isn't growing linearly with size of data) - you can do simple experiment by creating a lot of keys and then removing them. The size of data then is pretty large, which is not a very good sign.
2. at least for now, we won't be using it for monitoring
3. @piosz doesn't want to expose metrics that we won't be using...

I'm fine with leaving the code, but we shouldn't expose the metric at least.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+
+		// Run the etcdctl command and get the output.
+		if err := cmd.Start(); err != nil {
+			log.Fatal("Failed to start etcdctl command: ", err)
```
don't fatal (the same applies to all other places).

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+	// Map for storing last obtained request count for a given grpc request type.
+	lastCount := make(map[GRPCLabels]float64)
+
+	for {
```
The same comment as above - make this function only handle a single request, and another function that will be calling this one in a loop (and handling errors potentially)

shyamjvs replied:
Done. Thanks!

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+
+	for {
+		// Prepare the prom2json and jq commands to be executed in a pipe.
+		cmd1 := exec.Command("/bin/prom2json", etcdMetricsScrapeURI)
```
Make names more meaningful.

shyamjvs replied:
Not doing this `cmd1 | cmd2` thing anymore. Hope variable names elsewhere are meaningful.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		r, w := io.Pipe()
+		cmd1.Stdout = w
+		cmd2.Stdin = r
+		cmd2Out, _ := cmd2.StdoutPipe()
```
Why ignore error?

shyamjvs replied:
Fixed.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+		cmd2Out, _ := cmd2.StdoutPipe()
+
+		// Run the prom2json command.
+		if err := cmd1.Start(); err != nil {
```
Come on - we don't want to run some super complicated machinery like prom2json and rely on jq just to read metrics.

Reading metrics in prometheus format is pretty easy. We are doing this here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/metrics_util.go#L209

You need something pretty much the same - you are only interested in a different set of metrics

shyamjvs replied:
I was looking for something like this and got so close to making use of protobufs + parser, but fell back to using json as it seemed easier (in terms of time and effort).
Thanks a ton for pointing to this example, I'm using much simpler machinery now. :)

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,275 @@
+	// Serve our metrics on listenAddress/metricsPath.
+	log.Infoln("Listening on: ", listenAddress)
+	http.Handle(metricsPath, prometheus.Handler())
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
```
I don't think this is useful.

It would be much more useful to e.g. export when is the last time we updated each of the metrics.

shyamjvs replied:
I can remove this, but adding last update timestamp would bring in some extra code. As the collector threads would have to lock the timestamp variable and update it.
Also, timestamp would anyway be captured by time series for any prometheus server that scrapes this tool.

shyamjvs commented on cluster/images/etcd-version-monitor/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM 	quay.io/prometheus/busybox:latest
```
Changed it to scratch. Thanks for the suggestion.

shyamjvs commented on cluster/images/etcd-version-monitor/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM 	quay.io/prometheus/busybox:latest
+LABEL	maintainer	"Shyam JVS <shyamjvs@google.com>"
+
+COPY	./bin/etcdctl				/bin/etcdctl
```
Done.

shyamjvs commented:
@wojtek-t Fixed on comments. PTAL now.

xiang90 commented:
@shyamjvs 

Why do we need this tool? What is the use case?

shyamjvs commented:
@xiang90 It's for validating rollouts of etcd3 on existing clusters as we upgrade to 1.6. A simple tool for collecting and exporting etcd version related metrics.
Ref issue: https://github.com/kubernetes/kubernetes/pull/41080

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,279 @@
+
+// Initialize the prometheus instrumentation and client related flags.
+var (
+	listenAddress			string
```
Too many spaces/tabs.

shyamjvs replied:
Fixed.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,279 @@
+		return fmt.Errorf("Failed to finish etcdctl command: %v", err)
+	}
+
+	// Obtain the versioned data size from the comma-seperated output.
```
Please give an example of how the output looks like

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,279 @@
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("Failed to start etcdctl command: %v", err)
+	}
+	out, _ := ioutil.ReadAll(cmdOut)
```
Why ignoring error?

shyamjvs replied:
Fixed.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,279 @@
+	prometheus.Unregister(prometheus.NewGoCollector())	// Unregister go metrics.
+
+	// Spawn threads for periodically scraping etcd version metrics.
+	go getVersionPeriodically()
```
Please add stop channel to those.

shyamjvs replied:
Done. Had to change the last glog.Fatal() to Errorf() as we want the `defer close(stopCh)` to run before the program quits.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,279 @@
+	http.Handle(metricsPath, prometheus.Handler())
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(`<html>
+             <head><title>Etcd version metrics exporter</title></head>
```
I don't like using http with paragraphs etc - if you really want this, this should be understandable by tool.

shyamjvs replied:
Removed it as it serves no useful purpose.

shyamjvs commented:
@wojtek-t Fixed the comments. PTAL.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,293 @@
+	// The below metric is not actually registered as it is not a very accurate
+	// measure of the data size. Register this if needed in future. Also call the
+	// function 'getVersionedDataSizePeriodically()' if this metric is enabled.
+	etcdVersionedDataSize 	= prometheus.NewGaugeVec(
```
Please remove this metric completely - it's only misleading to have it defined and not registered.

shyamjvs replied:
Removed the metric and updated Makefile/Dockerfile/Readme too as we no longer need etcdctl.

shyamjvs commented:
@wojtek-t Fixed comments. PTAL.

fgrzadkowski commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,46 @@
+ARCH:=amd64
+GOLANG_VERSION?=1.7.4
+REGISTRY?=gcr.io/google_containers
+TAG?=test
```
Please change the tad to something like 0.1

shyamjvs replied:
Done.

fgrzadkowski commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,221 @@
+// Initialize the prometheus instrumentation and client related flags.
+var (
+	listenAddress		 string
+	metricsPath			 string
```
Does this pass golint?

shyamjvs replied:
I ran hack/verify-golint.sh and it passes successfully, so it seems like it should be ok. The convention for the spacing inside a var block doesn't seem to be fixed as this kind of spacing has been done at some places (for eg. https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/rbac/rest/storage_rbac.go#L80, https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/rbac/register.go#L40) and differently at other places (eg. https://github.com/kubernetes/kubernetes/blob/master/pkg/client/metrics/metrics.go#L39). I might be missing some style guide rule here, please let me know if so :)

wojtek-t replied:
No matter if it is passing gofmt or not, it looks so strange that you should fix it.
I already asked twice for it.

shyamjvs replied:
Sorry, forgot to reply to this comment. I had fixed it previously.

fgrzadkowski commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,221 @@
+			Name: "version_info_fetch_count",
+			Help: "Number of times etcd's version info was fetched, labeled by etcd's server binary and cluster version",
+		},
+		[]string{"server_version", "cluster_version"})
```
What is `server_version` and `cluster_version`? Which one refers to binary and which one to data format/API? I find it hard to distinguish, so maybe we should rename?

shyamjvs replied:
server_version is the version of the etcd's server (which is the binary). And cluster_version is the version of the API of the cluster, which is computed as min of all etcd API versions used in the cluster (@wojtek-t Can you confirm this?) Also, the names of these labels are the same as exposed by etcd, so just sticked to it for consistency.

wojtek-t replied:
Didn't we agreed an hour ago that we don't want to expose cluster_version at all?

And we want to rename "server_version" to "binary_version" or sth like that?
@fgrzadkowski ?

shyamjvs replied:
Changed it. Thanks!

fgrzadkowski commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,221 @@
+			Name: "grpc_requests_total",
+			Help: "Counter of received grpc requests, labeled by grpc method and grpc service names",
+		},
+		[]string{"grpc_method", "grpc_service"})
```
I think there is not need to repease "grpc" here. It's a label of "grpc" metrics, so it's rather obvious.

shyamjvs replied:
Again, this is something I sticked to for consistency with etcd. I felt it was weird too to stress on grpc twice, but that's how etcd exposes the metric.

wojtek-t replied:
+1 for what Filip wrote - please fix.

shyamjvs replied:
Done.

fgrzadkowski reviewed (commented):
I mostly checked metric definition, not the scraping code.

wojtek-t commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,46 @@
+ARCH:=amd64
+GOLANG_VERSION?=1.7.4
+REGISTRY?=gcr.io/google_containers
+TAG?=v0.1
```
to be consistent with naming we would like to have long term, change it to:
"v0.1.0"
(i.e. major, minor and patch release)

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,46 @@
+CURRENT_DIR:=$(pwd)
+TEMP_DIR:=$(shell mktemp -d)
+
+# Build the image for etcd-version-monitor.
```
Remove this comment - this doesn't bring any value.

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,46 @@
+	# Build the image.
+	docker build -t $(IMAGE) $(TEMP_DIR)
+
+# Push the etcd-version-monitor image to gcloud registry.
```
Remove this comment - this doesn't bring any value.

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/Makefile:
```diff
@@ -0,0 +1,46 @@
+	docker run -it -v $(shell pwd)/../../../:/go/src/k8s.io/kubernetes -v $(TEMP_DIR):/build -e GOARCH=$(ARCH) golang:$(GOLANG_VERSION) \
+		/bin/bash -c "CGO_ENABLED=0 go build -o /build/etcd-version-monitor k8s.io/kubernetes/cluster/images/etcd-version-monitor"
+
+	# Build the image.
```
Remove this comment - this doesn't bring any value.

shyamjvs replied:
Done.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,221 @@
+		if err := getVersion(); err != nil {
+			glog.Errorf("Failed to fetch etcd version: %v", err)
+		}
+		time.Sleep(scrapeTimeout)
```
I think we already talked about. This is pretty bad patter, as you cannot kill it when it is sleeping.
The pattern should be:
```
select {
case <-stopCh:
  break
case <-time.After(scrapeTimeout):
}
// real logic here
}
```

shyamjvs replied:
Done. Missed out this point, thanks.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,221 @@
+		if err := getGRPCRequestCount(&lastRecordedCount); err != nil {
+			glog.Errorf("Failed to fetch etcd grpc request counts: %v", err)
+		}
+		time.Sleep(scrapeTimeout)
```
Same as above.

shyamjvs replied:
Done.

shyamjvs commented:
@wojtek-t Fixed the comments. PTAL.

piosz commented:
@brancz @fabxc

piosz commented:
@brancz @fabxc can one of you please verify whether metrics are reasonable (follow up from our thread on slack and #41071)

brancz commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,219 @@
+}
+
+// Function for fetching etcd grpc request counts and feeding it to the prometheus metric.
+func getGRPCRequestCount(lastRecordedCount *map[GRPCRequestLabels]float64) error {
```
I'm failing to understand why this needs to be scraped from the etcd metrics endpoint and then exposed again. Why not scrape it directly from etcd? The version metrics should probably also be contributed to the upstream project, but I understand the possible need for it for all versions before that would be in an etcd release.

shyamjvs replied:
From an earlier discussion about this, it seemed better to have all version-related metrics (either the ones exposed directly by etcd or ones obtained through crunching (like the etcd version for example)) at one place. So we (re)expose some stuff by the version monitor. And yes, it would be great to have this upstream, but before that happens we need this for monitoring k8s cluster upgrades to 1.6.

brancz replied:
Sorry if I sounded too negative, I completely understand the need, just wanted to encourage to additionally get this into etcd itself :)

xiang90 replied:
@shyamjvs can you send a pr to upstream etcd to expose the version as a metrics? also you probably can write a simple prom query to get the aggregated number of request, no?

shyamjvs replied:
@fabxc We do not intend this to be a long-term monitoring system and its purpose for now is mainly to keep a watch on the upgrade. And even if we patch this logic to etcd to expose the version metric, it would still require replacing the current binaries in clusters and we don't want to take that risk without having a monitor. Chicken and egg! :)

@xiang90 Definitely all this logic should move into etcd and we do not want to need such a tool in the middle in long run. But we kind of need it as an urgency for the upcoming release.
> can you send a pr to upstream etcd to expose the version as a metrics?
I'll put this on my plate, but it'd need to wait for a while before I get done with some more urgent things :)
> also you probably can write a simple prom query to get the aggregated number of request
Assuming you are referring to gRPC requests, guess it's better to have them without aggregation as having more is better than less (unless I'm missing something you intended here).

fabxc commented:
If we do pull-proxy the gRPC request metrics, it should happen synchronously, i.e. when the prometheus handler gets called, it triggers a pull to the etcd `/metrics` endpoint.

The user configures their scrape interval against this tool and this should directly map to the scrape interval of the etcd-provided metrics. This also guarantees data freshness.

shyamjvs commented:
@wojtek-t @fgrzadkowski Could you PTAL once finally whether things seem fine?

wojtek-t reviewed (commented):
I'm roughly fine, modulo metrics definitions.

shyamjvs commented:
@wojtek-t @fgrzadkowski Thanks for the comments, fixed them. Additionally I made the following couple of changes:
- Deleting the old version metric if it's different from the new one (otherwise we don't touch the metric at all)
- Moved the `defer resp.Body.close()` lines after the error checking of receiving http response, otherwise the monitor crashes due to SEG fault if etcd instance isn't running.

PTAL

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,233 @@
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&listenAddress, "listen-address", ":9101", "Address to listen on for serving prometheus metrics")
```
Sorry - didn't catch this. This is pretty bad encapsulation.
This flag either should:
- be the full address (preferred)
- or just the port (that should be attached to localhost)

Having a port with ":" is super bad.

shyamjvs replied:
Done.

wojtek-t reviewed (commented):
One comment - other than that LGTM.

wojtek-t commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,233 @@
+
+// Initialize the prometheus instrumentation and client related flags.
+var (
+	listenAddress string
```
This doesn't pass gofmt - see bazel results.

shyamjvs replied:
Fixed :)

wojtek-t commented:
LGTM

@fgrzadkowski - do you also want to take a look?

fgrzadkowski commented:
I only checked metric names and they LGTM.

fgrzadkowski commented:
/lgtm

shyamjvs commented:
@wojtek-t Could you add approve label to the PR?

wojtek-t commented:
/lgtm
/approve

shyamjvs commented:
@wojtek-t Changed the image to `gcr.io/google-containers/etcd-version-monitor:0.1.0` (to be consistent with usage of hyphens instead of underscores in image names, like at other places) and updated the pod manifest yaml to the new image name. Please re-lgtm.

wojtek-t commented:
@shyamjvs - please fix tests first

shyamjvs commented:
@k8s-bot bazel test this
@k8s-bot kubemark e2e test this
@k8s-bot verify test this

shyamjvs commented:
@wojtek-t Fixed the verify test for flags. The kubemark and bazel failures seem unrelated.Thanks for pointing to it.

wojtek-t commented:
/lgtm

shyamjvs commented:
@wojtek-t Changed the gRPC requests count to a Gauge for code simplicity. PTAL.

shyamjvs commented:
Changed it back to counter on @wojtek-t 's request.

wojtek-t commented:
/lgtm

shyamjvs commented:
@k8s-bot cvm gce e2e test this

shyamjvs commented:
@k8s-bot kops aws e2e test this
@k8s-bot non-cri e2e test this

lavalamp commented on cluster/images/etcd-version-monitor/etcd-version-monitor.go:
```diff
@@ -0,0 +1,233 @@
+	scrapeTimeout        time.Duration
+)
+
+func registerFlags(fs *pflag.FlagSet) {
```
nit: if you're going to the trouble of registering these flags in a method like this, you might as well stuff the flags in a struct so that they don't have to be global.
